Add Filter patterns documentation (#1359)
hashicorp/design-system: website/docs/patterns/filter-patterns/partials/guidelines/guidelines.md

## Structure

A filtering pattern consists of a number of elements and Helios components that work together functionally and enforce a consistent layout. These elements are variable depending on the content the data set contains and the context within the application, and can be combined in different ways to achieve an experience that benefits the end-user.

### Data set

A data set is a broad term for an array items, objects, or related information presented as separate but related records. A data set is commonly represented via a [Table](/components/table) but may be expressed in other formats depending on the type of information.

**Dataset represented by a table**

![Tabular data set](/assets/patterns/filter-patterns/example-data-set-tabular-data.png)

**Dataset represented by grid**

![Non-tabular data set](/assets/patterns/filter-patterns/example-data-set-non-tabular-data.png)

### Filter positioning

> Filters are responsible for the functional aspect of limiting a data set and for displaying the [parameters](/patterns/filter-patterns?tab=core%20concepts#parameter) which a user can filter upon. They often consists of numerous parameters that mirror the columns of a table or metadata contained within the data set.

Reviewer: ```suggestion
Filters are responsible for the functional aspect of limiting a data set and for displaying the [parameters](/patterns/filter-patterns?tab=core%20concepts#parameter) which a user can filter upon. They often consist of numerous parameters that mirror the columns of a table or metadata contained within the data set.
```

> Filter positioning is largely dependent on the complexity of the data set and the scope for which the filters are being applied (are the filters applied at the page-level, or only to the data set they are paired with).

Reviewer: I believe there's no need to hyphenate here, but I might be wrong
```suggestion
Filter positioning is largely dependent on the complexity of the data set and the scope for which the filters are being applied (are the filters applied at the page level, or only to the data set they are paired with).
```

#### Filter bar

Organizing filters in a  **filter bar** is a common positioning method and consists of one or more dropdowns, buttons, or input components that are used to select values from a set of parameters. A [Segmented Group](/components/segmented-group) can be used within a filter bar to group similar parameters and support complex filtering.

![Filter bar example](/assets/patterns/filter-patterns/filter-bar-segmented-group.png)

**Ex 1.0: Horizontal filter bar**

![Horizontal orientation of the filter bar](/assets/patterns/filter-patterns/layout-filter-bar.png)

#### Sidebar

If space permits, filters can be positioned vertically in a **sidebar**, either at the page level, or in-line with the data set.

- **Page-level sidebar:** can be used for more complex data sets and if filtering contents of the entire page is required.
- **In-line sidebars:** can be used when the number of filterable parameters is minimal and if space permits.

**Ex 1.1 Page-level filter sidebar**

![Vertical orientation of the filter bar](/assets/patterns/filter-patterns/layout-sidebar-left-page-level.png)

**Ex 1.2 In-line filter sidebar**

> ![Vertical orientation of the filter bar on the left](/assets/patterns/filter-patterns/layout-sidebar-left-inline.png)

Reviewer: The Inline Sidebar demarcation here includes a part of the 'applied filters' area – I believe we can reduce the boundaries to only include the filters.
<img width="776" alt="Screenshot 2023-05-18 at 13 43 10" src="https://github.com/hashicorp/design-system/assets/788096/b6e2ebda-e6be-49e2-840a-676679fb0fc1">

![Applied filters](/assets/patterns/filter-patterns/applied-filters.png)

#### Positioning

Applied filters should be positioned as close as possible relative to the data set to help the user understand what has been filtered upon. In practice this is most often directly above the data set.

**Ex 2.0: Between the filter bar and the data set**

![Applied filters with filter bar positioning](/assets/patterns/filter-patterns/applied-filters-positioning-filter-bar.png)

**Ex 2.1: Above the data set with filters positioned in a sidebar**

![Applied filters with filter sidebar](/assets/patterns/filter-patterns/applied-filters-positioning-sidebar.png)

If the filters are positioned in a page-level sidebar the layout can be simplified by excluding the applied filters and including an action to clear all of the filters within the sidebar itself.

With this method, the filter values being applied to the dataset are already displayed and implied in the sidebar.

**Ex 2.2: Excludes the applied filters when positioned in a sidebar**

> ![Applied filters with clear all in the sidebar](/assets/patterns/filter-patterns/applied-filters-positioning-sidebar-clear-all.png)

Reviewer 2: Notice: here the icon of "clear all" is a plus instead of a cross, probably you want to use the same as in the other examples

### Pagination

> Use pagination to break down the filtered data set into pages. For more details refer to the [Pagination](/components/pagination) guidelines.

Reviewer 2: As a reader, I would have expected at least an illustrative image here, a visual example of a filter + table + pagination.

## Overflow

Depending on the complexity of the data set or filterable parameters, it may be necessary to account for overflow of elements within the filter bar or in the applied tags.

### Filter bar overflow

If the number of filterable parameters exceeds the available space, consider prioritizing the most commonly used filters in the filter bar and moving less important or more complex filters to a [Flyout](/components/flyout) that can be triggered via an action in the filter bar.

**Ex 3.0: Hiding additional filters behind an action; in this scenario a secondary button**

![More filters trigger](/assets/patterns/filter-patterns/overflow-filter-bar.png)

**Ex 3.1: Triggering a Flyout with additional filter parameters**

![Flyout triggered by more filters](/assets/patterns/filter-patterns/overflow-filter-bar-flyout.png)

!!! Dont

Don't stack multiple filter bars on top of a data set. This can result in unnecessary complexity in the UI.

![Multiple filter bar rows](/assets/patterns/filter-patterns/overflow-filter-bar-multiple-rows.png)

!!!

### Filter sidebar overflow

> Similar to the guidelines for the filter bar, an action to view more filters within a [Flyout](/components/flyout) can also be used triggered within the sidebar orientation.

Reviewer: This phrase doesn't read well (looking at 'used triggered') but not sure what was supposed to be.

Author: Yep, good catch

## Structure

A filtering pattern consists of a number of elements and Helios components that work together functionally and enforce a consistent layout. These elements are variable depending on the content of the data set and the context within the application, and can be combined in different ways to achieve an experience that benefits the end-user.

### Data set

A data set is a broad term for an array of items, objects, or related information presented as separate but related records. A data set is commonly represented via a [Table](/components/table) but may be expressed in other formats depending on the type of information.

**Dataset represented by a table**

![Tabular data set](/assets/patterns/filter-patterns/example-data-set-tabular-data.png)

**Dataset represented by a grid**

![Non-tabular data set](/assets/patterns/filter-patterns/example-data-set-non-tabular-data.png)

### Filter positioning

> Filters are responsible for the functional aspect of limiting a data set and for displaying the [parameters](/patterns/filter-patterns?tab=core%20concepts#parameter) which a user can filter upon. They often consist of numerous parameters that mirror the columns of a table or metadata contained within the data set.

Author: I separated it out and added some more context. I'm also covering this in the more conceptual docs: something like "mirroring the filters and the data set", when that's complete I can cross link between these areas of the docs.

> Filter positioning is largely dependent on the complexity of the data set and the scope for which the filters are being applied (are the filters applied at the page level, or only to the data set they are paired with).

Author: Not sure if you intended a `?` at the end of that suggestion?

> Filter positioning is largely dependent on the complexity of the data set and the scope for which the filters are being applied (are the filters applied at the page level, or only to the data set they are paired with).

Reviewer 3: When I ran a grammar check on this it came up with:

"Filter positioning is largely dependent on the complexity of the data set and the scope for which the filters are being applied (are the filters applied at the page level or only to the data set they are paired with?)."

> Filter positioning is largely dependent on the complexity of the data set and the scope for which the filters are being applied (are the filters applied at the page level, or only to the data set they are paired with).

Reviewer 3: When I ran a grammar check on this it came up with:

"Filter positioning is largely dependent on the complexity of the data set and the scope for which the filters are being applied (are the filters applied at the page level or only to the data set they are paired with?)."

#### Filter bar

Organizing filters in a  **filter bar** is a common positioning method and consists of one or more dropdowns, buttons, or input components that are used to select values from a set of parameters. A [Segmented Group](/components/segmented-group) can be used within a filter bar to group similar parameters and support complex filtering.

![Filter bar example](/assets/patterns/filter-patterns/filter-bar-segmented-group.png)

**Ex 1.0: Horizontal filter bar**

![Horizontal orientation of the filter bar](/assets/patterns/filter-patterns/layout-filter-bar.png)

#### Sidebar

If space permits, filters can be positioned vertically in a **sidebar**, either at the page level, or in-line with the data set.

- **Page-level sidebar:** can be used for more complex data sets and if filtering contents of the entire page is required.

> - **In-line sidebars:** can be used when the number of filterable parameters is minimal and if space permits.

Reviewer 3: ```suggestion
- **In-line sidebar:** can be used when the number of filterable parameters is minimal and if space permits.
```

**Ex 1.1: Page-level filter sidebar**

![Vertical orientation of the filter bar](/assets/patterns/filter-patterns/layout-sidebar-left-page-level.png)

**Ex 1.2: In-line filter sidebar**

![Vertical orientation of the filter bar on the left](/assets/patterns/filter-patterns/layout-sidebar-left-inline.png)

![Vertical orientation of the filter bar on the right](/assets/patterns/filter-patterns/layout-sidebar-right-inline.png)

### Applied filters

> The **applied filters** displays which filter values have been applied (if any) and provides a user with a method to clear or reset all filters at once. In most circumstances the filters that have been applied should be represented using a [Tag](/components/tag) component which allows for the individual dismissal of a filter value.

Reviewer 3: @jorytindall Was this meant to be resolved without applying the suggested changes? Just asking because there are a couple minor grammar and punctuation errors in the original.

> The **applied filters** displays which filter values have been applied (if any) and provides a user with a method to clear or reset all filters at once. In most circumstances the filters that have been applied should be represented using a [Tag](/components/tag) component which allows for the individual dismissal of a filter value.

Author: There are a bunch of changes that I've made on my local branch in aggregating the feedback, so I believe this branch is out of date. I'll push up the changes asap!

> ![Applied filters](/assets/patterns/filter-patterns/applied-filters.png)

Author: This actually uses a button (small variant). Functionally, that makes the most semantic sense and I believe that's what @alex-ju used in his exploration. If there's another route we want to go, let's discuss

> ![Applied filters](/assets/patterns/filter-patterns/applied-filters.png)

Reviewer: Indeed, we'd have to use a button for the 'clear filters' action – the closest to the design I could think of while exploring the implementation was a tertiary small button.

#### Positioning

> Applied filters should be positioned as close as possible relative to the data set to help the user understand what has been filtered upon. In practice this is most often directly above the data set.

Reviewer 3: ```suggestion
Applied filters should be positioned as close as possible to the data set to help the user understand what has been filtered upon. In practice, this is most often directly above the data set.
```

**Ex 2.0: Between the filter bar and the data set**

![Applied filters with filter bar positioning](/assets/patterns/filter-patterns/applied-filters-positioning-filter-bar.png)

**Ex 2.1: Above the data set with filters positioned in a sidebar**

![Applied filters with filter sidebar](/assets/patterns/filter-patterns/applied-filters-positioning-sidebar.png)

> If the filters are positioned in a page-level sidebar the layout can be simplified by excluding the applied filters and including an action to clear all of the filters within the sidebar itself.

Reviewer 3: ```suggestion
If the filters are positioned in a page-level sidebar, the layout can be simplified by excluding the applied filters and including an action to clear all of the filters within the sidebar itself.
```

With this method, the filter values being applied to the dataset are already displayed and implied in the sidebar.

**Ex 2.2: Excludes the applied filters when positioned in a sidebar**

> ![Applied filters with clear all in the sidebar](/assets/patterns/filter-patterns/applied-filters-positioning-sidebar-clear-all.png)

Author: I reorganized this and re-wrote it a bit.

### Pagination

> Use pagination to break down the filtered data set into pages. For more details refer to the [Pagination](/components/pagination) guidelines.

Reviewer 3: ```suggestion
Use pagination to break down the filtered data set into pages. For more details, refer to the [Pagination](/components/pagination) guidelines.
```

> ![Pagination example](/assets/patterns/filter-patterns/pagination-layout.png)

Reviewer 4: [nit] The casing of "Applied Filters" is different than the casing of the other sections in this example. It should be "Applied filters"

> ## Overflow

Reviewer 4: I think this portion of the pattern could have large usability implications, and I'm concerned about moving forward with these recommended approaches without first validating them in the products. I wonder if we should indicate to consumers that these are untested approaches and that they should share any user feedback they receive with the HDS team to help us improve this pattern.

> Depending on the complexity of the data set or filterable parameters, it may be necessary to account for overflow of elements within the filter bar or in the applied tags.

Reviewer 4: [nit] less wordy
```suggestion
Depending on the complexity of the data set, it may be necessary to account for the overflow of elements within the filter bar or in the applied tags.
```

### Filter bar overflow

If the number of filterable parameters exceeds the available space, consider prioritizing the most commonly used filters in the filter bar and moving less important or more complex filters to a [Flyout](/components/flyout) that can be triggered via an action in the filter bar.

**Ex 3.0: Hiding additional filters behind an action; in this scenario a secondary button**

> ![More filters trigger](/assets/patterns/filter-patterns/overflow-filter-bar.png)

Author: Strategically updating these as I go since the images are spread across multiple PRs

**Ex 3.1: Triggering a Flyout with additional filter parameters**

![Flyout triggered by more filters](/assets/patterns/filter-patterns/overflow-filter-bar-flyout.png)

!!! Dont

Don't stack multiple filter bars on top of a data set. This can result in unnecessary complexity in the UI.

![Multiple filter bar rows](/assets/patterns/filter-patterns/overflow-filter-bar-multiple-rows.png)

> !!!

Reviewer 4: I think this would be more impactful if it weren't completely out of the realm of possibility. For example, it might be good to showcase an example we've seen recently of a more complex filtering variation, just done incorrectly. This "don't" example is really supposed to showcase stacking filters, no multiple search fields. It just feels like it's doing too much.

### Filter sidebar overflow

Similar to the guidelines for the filter bar, an action to view more filters within a [Flyout](/components/flyout) can be triggered within the sidebar orientation.

**Ex 3.2: Hiding additional filters behind an action in a sidebar orientation**

> ![More filters trigger in a sidebar](/assets/patterns/filter-patterns/overflow-sidebar-more-filters.png)

Reviewer 4: Why wouldn't this result in a scrollable area? The flyout use case doesn't make sense to me here because the scrollable real estate could be used here.

!!! Info

Consider each one of these scenarios carefully as moving additional filters to an interstitial component like a Flyout adds extra steps in the filtering process that might not be necessary. 

!!!

### Applied filters overflow

> If the number of filterable parameters is large, there's a likelihood that the number of applied filters may also extend to multiple lines. In this scenario the same methods apply as the filter bar and sidebar.

Reviewer 4: > In this scenario the same methods apply as the filter bar and sidebar.

Have you thought about something more like a show/hide that expands the FilterList instead of opening a Flyout to show the other filters? As you noted, these are here to help with easy and quick dismissal, so why would we hide them within a Flyout?

Author: My intention here (and with the overflow content in general) was to only use components/patterns/things are the currently supported within Helios. We briefly discussed whether we should recommend a reveal/show-hide based on the work in progress, but since it isn't complete yet decided against.

> If the number of filterable parameters is large, there's a likelihood that the number of applied filters may also extend to multiple lines. In this scenario the same methods apply as the filter bar and sidebar.

Reviewer 3: ```suggestion
If the number of filterable parameters is large, there’s a likelihood that the number of applied filters may also extend to multiple lines. In this scenario the same methods apply as the filter bar and sidebar.
```

Filters are responsible for the functional aspect of limiting a data set and for displaying the [parameters](/patterns/filter-patterns?tab=core%20concepts#parameter) which a user can filter upon. Filter positioning is largely dependent on the complexity of the data set and the scope to which the filters are being applied. For example, are the filters applied at the page level or only to the data set they are paired with?

Filters often consist of numerous parameters that mirror the columns of a table or parameters contained within the data set. For more details on mirroring filters and parameters within a data set, refer to the filtering [core concepts](/patterns/filter-patterns/tab=core%20concepts#mirror-of-parameters).

#### Filter bar

Organizing filters in a **filter bar** is a common positioning method and consists of one or more dropdowns, buttons, or input components that are used to select values from a set of parameters. A [Segmented Group](/components/segmented-group) can be used within a filter bar to group similar parameters and support complex filtering.

A filter bar should be positioned directly above the data set with a a `16px` gap in between.

> ![Filter bar example](/assets/patterns/filter-patterns/filter-bar-segmented-group.png =728x*)

Reviewer 5: (suggestion) you'll want to think about the alt attribute text. The guidelines indicate that the alt attribute value should convey what the image conveys- no more, no less.

Author: Yep, good point I'll do a scrub of all of the images

### Filters

Filters are responsible for the functional aspect of limiting a data set and for displaying the [parameters](/patterns/filter-patterns?tab=core%20concepts#parameter) which a user can filter upon. Filter positioning is largely dependent on the complexity of the data set and the scope to which the filters are being applied. For example, are the filters applied at the page level or only to the data set they are paired with?

Filters often consist of numerous parameters that mirror the columns of a table or parameters contained within the data set. For more details on mirroring filters and parameters within a data set, refer to the filtering [core concepts](/patterns/filter-patterns/tab=core%20concepts#mirror-of-parameters).

#### Filter bar

Organizing filters in a **filter bar** is a common positioning method and consists of one or more dropdowns, buttons, or input components that are used to select values from a set of parameters. A [Segmented Group](/components/segmented-group) can be used within a filter bar to group similar parameters and support complex filtering.

> ![Filters represented by a filter bar](/assets/patterns/filter-patterns/filter-bar-segmented-group.png =728x*)

Reviewer 4: ```suggestion
![Filters represented by a filter bar](/assets/patterns/filter-patterns/filter-bar-segmented-group.png =724x*)
```

### Applied filters

Display **applied filters** and provide the user with a method to clear all filters at once. Applied filters should be represented using a [Tag](/components/tag) component, which allows for the individual dismissal of a filter value.

![Applied filters](/assets/patterns/filter-patterns/applied-filters.png =450x*)

#### Positioning with filter bars

Applied filters should be positioned between the data set and the filter bar with a `16px` gap between elements.

![Applied filters with filter bar positioning](/assets/patterns/filter-patterns/applied-filters-positioning-filter-bar.png =559x*)

In the case of a filter sidebar, the applied filters should be positioned directly above the dataset with a `16px` gap.

![Applied filters with filter sidebar](/assets/patterns/filter-patterns/applied-filters-positioning-sidebar.png =559x*)

If the filters are positioned in a page-level sidebar, sometimes displaying the applied filters isn't necessary because they are implied in the sidebar. An action to clear all of the filters should be included within the sidebar itself.

> ![Applied filters with clear all in the sidebar](/assets/patterns/filter-patterns/applied-filters-positioning-sidebar-clear-all.png =559x*)

Reviewer 4: This section doesn't feel helpful for a consumer because these contradict one another. In the first example, you mention that they should show the applied filters above the dataset with a clear button at the end. In the second example, you mention that sometimes that isn't necessary, and the clear button should be in the sidebar. 

Let's lean into external research here and make an opinionated decision. We recommend that they either always show them or don't show them and then clarify why. We can always revise that decision if/when we get user feedback.

Author: Ok the opinionated decision here is to always display them above the data set; this is currently what Developer does and I think it's a the most explicit method to 1) maintain consistency across the different filter orientations and 2) maintain consistency with the single filter clear and the bulk clear.

### Global filter functions

Depending on the fetching method and complexity of the data set, global filter functions can be used to expose methods that effect the entire data set. Global functions should be differentiated from the primary filters and aligned to the end of the data set. Examples include:

- Manually refreshing the data set
- Triggering a [Flyout](/components/flyout) with more complex filters

#### Multiple global filter functions

![Multiple global functions](/assets/patterns/filter-patterns/global-filter-functions.png =874x*)

#### Single global function

![Single function](/assets/patterns/filter-patterns/global-filter-functions-single.png =874x*)

In the case a multiple filter functions, use the [Segmented Group](/components/segmented-group). If only a single function is necessary, use the secondary [Button](/components/button#secondary).

### Pagination

Use pagination to break down the filtered data set into pages. For more details, refer to the [Pagination](/components/pagination) guidelines.

![Pagination example](/assets/patterns/filter-patterns/pagination-layout.png =559x*)

> ## Empty state

Reviewer 4: The "Empty state" section feels less important than filtering methods and how to display results. I recommend moving it down after the "Displaying selected filters" section instead.

Also, this doesn't need to be addressed in this iteration, but I'd like us to make a note that we should add documentation to help designers create UIs that actively help users avoid an empty state. One way to do that is to show the # of results available for a particular value, which is shown in some of your examples and the videos, but isn't documented.

Use the Helios [Application State](/components/application-state) component to communicate the lack of results and steps the user can take to remedy the situation.

![Empty state example](/assets/patterns/filter-patterns/filter-patterns-empty-state.png =559x*)

### Within applied filters

If no filters have been applied, it can be helpful to communicate this in the applied filters area. Use a [Tooltip](/components/tooltip) coupled with the applied filters label to guide the user to the filter functions.

![Empty state in applied filters](/assets/patterns/filter-patterns/applied-filters-empty-state.png =450x*)

## Filtering methods

Depending on the fetching strategy, size, and rendering method of the data set, how filters are applied and the impact on the user experience varies.

### Live filtering

Live filtering refers to updating the records within a data set immediately after a user makes a selection. This method results in a responsive user experience, but may not be possible with large data sets or complex filtering methods.

<video width="100%" controls loop>
  <source
    src="/assets/patterns/filter-patterns/live-filtering.mp4"
    type="video/mp4"
  />

> </video>

Reviewer 4: In the "Empty state" section, you showcase the empty state on applied filters with a tooltip, but that's missing from the video. If that's our recommended approach, then our examples should also showcase that.

Reviewer 4: It looks like all the videos are that way, actually.

Author: Updated this!

**Communicating that multiple filter parameters have been applied**

![Filter dropdown with multiple counts](/assets/patterns/filter-patterns/filter-bar-dropdown-count-multiple.png =683x*)

## Overflow

Depending on the complexity of the data set, it may be necessary to account for the overflow of elements within the filter bar or in the applied tags.

### Filter bar overflow

If the number of filterable parameters exceeds the available space, consider prioritizing the most commonly used filters in the filter bar and moving less important or more complex filters to a [Flyout](/components/flyout) that can be triggered via an action in the filter bar.

**Hiding additional filters behind an action; in this scenario a secondary button**

![More filters trigger](/assets/patterns/filter-patterns/overflow-filter-bar.png =559x*)

**Triggering a Flyout with additional filter parameters**

![Flyout triggered by more filters](/assets/patterns/filter-patterns/overflow-filter-bar-flyout.png =559x*)

!!! Dont

Don't stack multiple filter bars on top of a data set. This can result in unnecessary complexity in the UI.

> ![Multiple filter bar rows](/assets/patterns/filter-patterns/overflow-filter-bar-multiple-rows.png =559x*)

Reviewer 4: It would be more impactful to use a real example here.

!!!

### Filter sidebar overflow

Similar to the guidelines for the filter bar, an action to view more filters within a [Flyout](/components/flyout) can be triggered within the sidebar orientation.

**Hiding additional filters behind an action in a sidebar orientation**

![More filters trigger in a sidebar](/assets/patterns/filter-patterns/overflow-sidebar-more-filters.png =559x*)

!!! Info

Consider each one of these scenarios carefully as moving additional filters to an interstitial component like a Flyout adds extra steps in the filtering process that might not be necessary.

> !!!

Reviewer 4: Can you help me understand why this wouldn't just scroll to reveal more filters?

Author: Updated this!

### Applied filters overflow

If the number of filterable parameters is large, there’s a likelihood that the number of applied filters may also extend to multiple lines. In this scenario the same methods apply as the filter bar and sidebar.

Depending on the context of the filtering pattern, it may be helpful to display the applied filters to the user at all times. This can assist in easy dismissal of specific filters and help the user better parse how a data set is being filtered.

**Extreme example of many applied filters**

![Many filters applied](/assets/patterns/filter-patterns/overflow-applied-filters.png =559x*)

**Applied filters within a Flyout**

> ![Applied filters within a Flyout](/assets/patterns/filter-patterns/overflow-applied-filters-flyout.png =559x*)

Reviewer 4: This interaction could be simplified. Have you considered using a "show more" action that reveals the rest of the filters in place instead of opening a flyout? 

This still feels a little like a "choose your own adventure," and we should also be opinionated here. 
Let's pick one and recommend that approach to work towards consistency across the UIs:
- Applied filters are always all visible.
- Only show a certain number of applied filters and then an action to display them in a Flyout (I do not recommend this approach; it feels unnecessarily complex and uncommon).
- Only show a certain number of applied filters and then reveal the rest in place via a "show more" action.
